File: scripts/extract-budget.py
"""
extract-budget.py
Reads "BUDGET (1).xlsx" and writes src/data/budget.json

Run from project root:
    python3 scripts/extract-budget.py

Install dep if needed:
    pip install openpyxl
"""

import json, os
from pathlib import Path
import openpyxl

ROOT = Path(__file__).parent.parent
wb = openpyxl.load_workbook(ROOT / "BUDGET (1).xlsx", data_only=True)

# Use the most recent month sheet (skip CARD USED - LIST)
month_sheets = [s for s in wb.sheetnames if s != "CARD USED - LIST"]
sheet_name = month_sheets[-1]
ws = wb[sheet_name]

rows = list(ws.iter_rows(values_only=True))

# ── Income ──────────────────────────────────────────────────────────────────
total_income = 0
raw_free_spend = 0

for row in rows:
    if row[0] == "Total" and isinstance(row[1], (int, float)):
        total_income = float(row[1])
    if isinstance(row[0], str) and "raw free spend" in row[0].lower():
        raw_free_spend = float(row[1]) if isinstance(row[1], (int, float)) else 0

# ── Expenses ─────────────────────────────────────────────────────────────────
FIXED_LABELS = {
    "Rent", "Roth IRA", "Subscriptions", "Phone Bill", "Gym Membership",
    "Internet", "MTA", "Savings", "Seperate Account",
}
VARIABLE_LABELS = {
    "Groceries", "Eating out", "Coffee/ drinks", "Ubers",
    "Fun", "Travel Fund", "Emergency Savings", "Investing", "misc", "claude",
}

fixed_expenses = []
variable_expenses = []
in_fixed = False
in_variable = False

for row in rows:
    if row[0] == "Fixed Expenses":
        in_fixed, in_variable = True, False
        continue
    if row[0] == "Variable Expenses":
        in_fixed, in_variable = False, True
        continue
    if in_fixed and row[0] in FIXED_LABELS:
        fixed_expenses.append({
            "label": row[0],
            "budget": float(row[1] or 0),
            "actual": float(row[2] or 0),
        })
    if in_variable and row[0] in VARIABLE_LABELS:
        variable_expenses.append({
            "label": row[0],
            "budget": float(row[1] or 0),
            "actual": float(row[2] or 0),
        })

# ── Savings ──────────────────────────────────────────────────────────────────
savings_row = next((r for r in fixed_expenses if r["label"] == "Savings"), None)
separate_row = next((r for r in fixed_expenses if r["label"] == "Seperate Account"), None)
savings_target = (savings_row["budget"] if savings_row else 0) + (separate_row["budget"] if separate_row else 0)
savings_actual = (savings_row["actual"] if savings_row else 0) + (separate_row["actual"] if separate_row else 0)

# ── Free spend ───────────────────────────────────────────────────────────────
variable_actual_spent = sum(r["actual"] for r in variable_expenses)
free_spend_remaining = raw_free_spend - variable_actual_spent

output = {
    "month": sheet_name,
    "totalIncome": total_income,
    "rawFreeSpend": raw_free_spend,
    "freeSpendRemaining": free_spend_remaining,
    "variableActualSpent": variable_actual_spent,
    "savingsTarget": savings_target,
    "savingsActual": savings_actual,
    "fixedExpenses": fixed_expenses,
    "variableExpenses": variable_expenses,
}

out_path = ROOT / "src" / "data" / "budget.json"
out_path.parent.mkdir(parents=True, exist_ok=True)
out_path.write_text(json.dumps(output, indent=2))

print(f"✅  Budget extracted from '{sheet_name}' sheet")
print(f"    Income:        ${total_income:,.2f}")
print(f"    Free to spend: ${free_spend_remaining:,.2f} / ${raw_free_spend:,.2f}")
print(f"    Savings:       ${savings_actual:,.2f} / ${savings_target:,.2f}")
print(f"    Written to:    src/data/budget.json")

File: scripts/extract-budget.mjs
/**
 * extract-budget.mjs
 * Reads BUDGET (1).xlsx and writes src/data/budget.json
 * Run with: node scripts/extract-budget.mjs
 *
 * Install dep if needed: npm install xlsx
 */

import { readFile, writeFile, mkdir } from "fs/promises";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const root = join(__dirname, "..");

// Dynamically import xlsx (works with both ESM and CJS builds)
const XLSX = await import("xlsx").then(m => m.default ?? m);

const wb = XLSX.readFile(join(root, "BUDGET (1).xlsx"));

// Find the most recent month sheet (last non-CARD sheet)
const monthSheets = wb.SheetNames.filter(n => n !== "CARD USED - LIST");
const sheetName = monthSheets[monthSheets.length - 1]; // e.g. "June"
const ws = wb.Sheets[sheetName];
const rows = XLSX.utils.sheet_to_json(ws, { header: 1, defval: null });

// ── Extract income ──────────────────────────────────────────────────────────
let totalIncome = 0;
let rawFreeSpend = 0;

for (const row of rows) {
  if (row[0] === "Total" && typeof row[1] === "number") {
    totalIncome = row[1];
  }
  if (typeof row[0] === "string" && row[0].toLowerCase().includes("raw free spend")) {
    rawFreeSpend = typeof row[1] === "number" ? row[1] : 0;
  }
}

// ── Extract fixed expenses ──────────────────────────────────────────────────
const FIXED_LABELS = [
  "Rent", "Roth IRA", "Subscriptions", "Phone Bill", "Gym Membership",
  "Internet", "MTA", "Savings", "Seperate Account",
];
const fixedExpenses = [];
let inFixed = false;
let inVariable = false;

const variableExpenses = [];
const VARIABLE_LABELS = [
  "Groceries", "Eating out", "Coffee/ drinks", "Ubers",
  "Fun", "Travel Fund", "Emergency Savings", "Investing", "misc", "claude",
];

for (const row of rows) {
  if (row[0] === "Fixed Expenses") { inFixed = true; inVariable = false; continue; }
  if (row[0] === "Variable Expenses") { inFixed = false; inVariable = true; continue; }

  if (inFixed && FIXED_LABELS.includes(row[0])) {
    fixedExpenses.push({ label: row[0], budget: row[1] ?? 0, actual: row[2] ?? 0 });
  }
  if (inVariable && VARIABLE_LABELS.includes(row[0])) {
    variableExpenses.push({ label: row[0], budget: row[1] ?? 0, actual: row[2] ?? 0 });
  }
}

// ── Savings target & actual ─────────────────────────────────────────────────
const savingsRow = fixedExpenses.find(r => r.label === "Savings");
const separateRow = fixedExpenses.find(r => r.label === "Seperate Account");
const savingsTarget = (savingsRow?.budget ?? 0) + (separateRow?.budget ?? 0);
const savingsActual = (savingsRow?.actual ?? 0) + (separateRow?.actual ?? 0);

// ── Total variable actual spent ─────────────────────────────────────────────
const variableActualSpent = variableExpenses.reduce((s, r) => s + (r.actual ?? 0), 0);
const freeSpendRemaining = rawFreeSpend - variableActualSpent;

// ── Output ──────────────────────────────────────────────────────────────────
const output = {
  month: sheetName,
  extractedAt: new Date().toISOString(),
  totalIncome,
  rawFreeSpend,
  freeSpendRemaining,
  variableActualSpent,
  savingsTarget,
  savingsActual,
  fixedExpenses,
  variableExpenses,
};

await mkdir(join(root, "src/data"), { recursive: true });
await writeFile(join(root, "src/data/budget.json"), JSON.stringify(output, null, 2));

console.log(`✅ Budget extracted from "${sheetName}" sheet`);
console.log(`   Income: $${totalIncome.toLocaleString()}`);
console.log(`   Free to spend: $${freeSpendRemaining.toLocaleString()} / $${rawFreeSpend.toLocaleString()}`);
console.log(`   Savings: $${savingsActual.toLocaleString()} / $${savingsTarget.toLocaleString()}`);
console.log(`   Written to src/data/budget.json`);
